File: [8] String to Integer (atoi)/8.string-to-integer-atoi.java
/*
 * @lc app=leetcode id=8 lang=java
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
class Solution {
    public int myAtoi(String s) {
        s = s.trim();
        if (s.isEmpty()) return 0;
        int i = 0, ans = 0, sign = 1, len = s.length();
		if (s.charAt(i) == '-' || s.charAt(i) == '+')
			sign = s.charAt(i++) == '+' ? 1 : -1;
		for (; i < len; i++) {
			int tmp = s.charAt(i) - '0';
			if (tmp < 0 || tmp > 9)
				break;
			if (ans > Integer.MAX_VALUE / 10
					|| (ans == Integer.MAX_VALUE / 10 && Integer.MAX_VALUE % 10 < tmp))
				return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
			else
				ans = ans * 10 + tmp;
		}
		return sign * ans;
    }
}
// @lc code=end

File: [8] String to Integer (atoi)/8.string-to-integer-atoi.cpp
/*
 * @lc app=leetcode id=8 lang=cpp
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start
class Solution {
public:
    int myAtoi(string s) {
        if (s.length() == 0) return 0;
        int i = 0;
        while (s[i] == ' ') i++;
        if (s.length() == i) return 0;
        int ans = 0, sign = 1;
        if (s[i] == '+' || s[i] == '-') {
            sign = (s[i] == '+')? 1: -1;
            i++;
        }
        while (s[i] >= '0' && s[i] <= '9') {
            int temp = s[i] - '0';
            if (ans > INT_MAX/10 || (ans == INT_MAX/10 && temp > INT_MAX%10)) {
                return (sign == 1)? INT_MAX: INT_MIN;
            } else {
                ans = ans*10+temp;
                i++;
            }
        }
        return sign*ans;
    }
};
// @lc code=end

